# Web Scraping Lab

You will find in this notebook some scrapy exercises to practise your scraping skills.

**Tips:**

- Check the response status code for each request to ensure you have obtained the intended contennt.
- Print the response text in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.

- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Beautiful Soup Doc](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [Urllib](https://docs.python.org/3/library/urllib.html#module-urllib)
- [re lib](https://docs.python.org/3/library/re.html)
- [lxml lib](https://lxml.de/)
- [Scrapy](https://scrapy.org/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### Below are the libraries and modules you may need. `requests`,  `BeautifulSoup` and `pandas` are imported for you. If you prefer to use additional libraries feel free to uncomment them.

In [3]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
#from dotenv import load_dotenv
#import os
# from pprint import pprint
# from lxml import html
# from lxml.html import fromstring
# import urllib.request
# from urllib.request import urlopen
# import random
import re
# import scrapy

#### Download, parse (using BeautifulSoup), and print the content from the Trending Developers page from GitHub:

In [2]:
# This is the url you will scrape in this exercise
url = "https://github.com/trending/developers"
response = requests.get(url)
print(response.status_code)
print(response)
print()
data=response.text
soup = BeautifulSoup(data,'html.parser')
print(soup.prettify())

200
<Response [200]>

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <link href="https://github.githubassets.com" rel="dns-prefetch"/>
  <link href="https://avatars0.githubusercontent.com" rel="dns-prefetch"/>
  <link href="https://avatars1.githubusercontent.com" rel="dns-prefetch"/>
  <link href="https://avatars2.githubusercontent.com" rel="dns-prefetch"/>
  <link href="https://avatars3.githubusercontent.com" rel="dns-prefetch"/>
  <link href="https://github-cloud.s3.amazonaws.com" rel="dns-prefetch"/>
  <link href="https://user-images.githubusercontent.com/" rel="dns-prefetch"/>
  <link crossorigin="anonymous" href="https://github.githubassets.com/assets/frameworks-6b8b7859c4b8fbe3ab45f8ab0905a9f8.css" integrity="sha512-aVn2DoCuXdXX9G3sp/Luupl/Ui00/iXrUh7Ke3geLlkigQY8GHBky7kKRSuyeKxGApWDdCQUy+6gTF1ZmYHWkw==" media="all" rel="stylesheet">
   <link crossorigin="anonymous" href="https://github.githubassets.com/assets/site-b046b27487428b94fc20941868838997.css" integ

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names. You can achieve this using Chrome DevTools.

1. Use BeautifulSoup to extract all the html elements that contain the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below:

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [3]:
developers=[e.text for e in soup.find_all(class_="h3 lh-condensed")]
for developer in developers:
    print(developer)

Vítor Galvão
Marcel Pociot
Luke Latham
Francois Zaninotto
Paulus Schoutsen
Vadim Demedes
Raphaël Benitte
John Otander
Mercy Ma
Hong Phuc Dang
Hisham Muhammad
Daniel Wirtz
Magnus Edenhill
Mike Perham
Daniel Martí
Kevin Papst
Thomas Lin Pedersen
Rakan Nimer
Raphaël Huchet
Andrew Gallant
Sheng Chen
Harry
Sean McArthur
Rob Dodson
Andrew Arnott


#### Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [4]:
#This is the url you will scrape in this exercise
url = 'https://github.com/trending/python?since=daily'

In [5]:
response = requests.get(url)
print(response.status_code)
print(response)
print()
data=response.text
soup = BeautifulSoup(data,'html.parser')
e=soup.find_all(class_="h3 lh-condensed")
for a in e:
    print(a.find("a").get("href")[1:])

200
<Response [200]>

testerSunshine/12306
Yorko/mlcourse.ai
iperov/DeepFaceLab
ckiplab/ckiptagger
tensorflow/neural-structured-learning
insilicomedicine/GENTRL
hudson-and-thames/mlfinlab
deepfakes/faceswap
CorentinJ/Real-Time-Voice-Cloning
deepinsight/insightface
zhaipro/easy12306
keras-team/keras
mne-tools/mne-python
pjialin/py12306
xadrianzetx/fullstack.ai
ansible/awx
s3nh/pytorch-text-recognition
alexmojaki/heartrate
RaRe-Technologies/gensim
open-mmlab/mmdetection
tensorflow/models
Azure/azure-cli
nprapps/heat-income
zalandoresearch/flair
iGhibli/iOS-DeviceSupport


#### Display all the image links from Walt Disney wikipedia page

In [6]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [7]:
response = requests.get(url)
print(response.status_code)
print(response)
print()
data=response.text
soup = BeautifulSoup(data,'html.parser')
result=["https://en.wikipedia.org" + link.get('href') for link in soup.find_all('a',class_="image")]
for element in result:
    print(element)

200
<Response [200]>

https://en.wikipedia.org/wiki/File:Walt_Disney_1946.JPG
https://en.wikipedia.org/wiki/File:Walt_Disney_1942_signature.svg
https://en.wikipedia.org/wiki/File:Walt_Disney_envelope_ca._1921.jpg
https://en.wikipedia.org/wiki/File:Trolley_Troubles_poster.jpg
https://en.wikipedia.org/wiki/File:Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg
https://en.wikipedia.org/wiki/File:Steamboat-willie.jpg
https://en.wikipedia.org/wiki/File:Walt_Disney_1935.jpg
https://en.wikipedia.org/wiki/File:Walt_Disney_Snow_white_1937_trailer_screenshot_(13).jpg
https://en.wikipedia.org/wiki/File:Disney_drawing_goofy.jpg
https://en.wikipedia.org/wiki/File:DisneySchiphol1951.jpg
https://en.wikipedia.org/wiki/File:WaltDisneyplansDisneylandDec1954.jpg
https://en.wikipedia.org/wiki/File:Walt_disney_portrait_right.jpg
https://en.wikipedia.org/wiki/File:Walt_Disney_Grave.JPG
https://en.wikipedia.org/wiki/File:Roy_O._Disney_with_Company_at_Press_Confer

#### Retrieve an arbitary Wikipedia page of "Python" and create a list of links on that page

In [8]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 

In [9]:
response = requests.get(url)
print(response.status_code)
print(response)
print()
data=response.text
soup = BeautifulSoup(data,'html.parser')
#"https://en.wikipedia.org"
result=["https://en.wikipedia.org" + str(link.get('href')) for link in soup.find_all('a')]
for element in result:
    print(element)


200
<Response [200]>

https://en.wikipedia.orgNone
https://en.wikipedia.org#mw-head
https://en.wikipedia.org#p-search
https://en.wikipedia.orghttps://en.wiktionary.org/wiki/Python
https://en.wikipedia.orghttps://en.wiktionary.org/wiki/python
https://en.wikipedia.org#Snakes
https://en.wikipedia.org#Ancient_Greece
https://en.wikipedia.org#Media_and_entertainment
https://en.wikipedia.org#Computing
https://en.wikipedia.org#Engineering
https://en.wikipedia.org#Roller_coasters
https://en.wikipedia.org#Vehicles
https://en.wikipedia.org#Weaponry
https://en.wikipedia.org#People
https://en.wikipedia.org#Other_uses
https://en.wikipedia.org#See_also
https://en.wikipedia.org/w/index.php?title=Python&action=edit&section=1
https://en.wikipedia.org/wiki/Pythonidae
https://en.wikipedia.org/wiki/Python_(genus)
https://en.wikipedia.org/w/index.php?title=Python&action=edit&section=2
https://en.wikipedia.org/wiki/Python_(mythology)
https://en.wikipedia.org/wiki/Python_of_Aenus
https://en.wikipedia.org/wiki

#### Number of Titles that have changed in the United States Code since its last release point 

In [10]:
# This is the url you will scrape in this exercise
url = 'https://uscode.house.gov/download/download.shtml'

In [11]:
#your code
response = requests.get(url)
print(response.status_code)
print(response)
data=response.text
soup = BeautifulSoup(data,'html.parser')
result=[re.sub("\n", "", re.sub("   *", "", e.text)) for e in soup.select('[id^="us/usc/"]')]
print()
for element in result:
    print(element)
print()
print(str(len(result))+" titles have changed, including Appendixes")


200
<Response [200]>

Title 1 - General Provisions ٭
Title 2 - The Congress
Title 3 - The President ٭
Title 4 - Flag and Seal, Seat of Government, and the States ٭
Title 5 - Government Organization and Employees ٭
Appendix
Title 6 - Domestic Security
Title 7 - Agriculture
Title 8 - Aliens and Nationality
Title 9 - Arbitration ٭
Title 10 - Armed Forces ٭
Title 11 - Bankruptcy ٭
Appendix
Title 12 - Banks and Banking
Title 13 - Census ٭
Title 14 - Coast Guard ٭
Title 15 - Commerce and Trade
Title 16 - Conservation
Title 17 - Copyrights ٭
Title 18 - Crimes and Criminal Procedure ٭
Appendix
Title 19 - Customs Duties
Title 20 - Education
Title 21 - Food and Drugs
Title 22 - Foreign Relations and Intercourse
Title 23 - Highways ٭
Title 24 - Hospitals and Asylums
Title 25 - Indians
Title 26 - Internal Revenue Code
Title 27 - Intoxicating Liquors
Title 28 - Judiciary and Judicial Procedure ٭
Appendix
Title 29 - Labor
Title 30 - Mineral Lands and Mining
Title 31 - Money and Finance ٭
Title 32 - 

#### A Python list with the top ten FBI's Most Wanted names 

In [139]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'

In [140]:
response = requests.get(url)
print(response.status_code)
print(response)
print()
data=response.text
soup = BeautifulSoup(data,'html.parser')
result=[e.text.replace("\n","").capitalize() for e in soup.select('h3',class_="title")]
for element in result:
    print(element)

200
<Response [200]>

Yaser abdel said
Jason derek brown
Alexis flores
Eugene palmer
Santiago villalba mederos
Rafael caro-quintero
Robert william fisher
Bhadreshkumar chetanbhai patel
Arnoldo jimenez
Alejandro rosales castillo


####  20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [141]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'

In [166]:
#your code
response = requests.get(url)
print(response.status_code)
print(response)
print()
soup = BeautifulSoup(response.text.encode('utf-8'))
table=soup.select("tbody tr")
dateColumn=[]
timeColumn=[]
latituteColumn=[]
longitudeColumn=[]
regionColumn=[]

for e in table:
    date,_=e.select("td")[3].find("a").text.split("\xa0\xa0\xa0")
    _,time=e.select("td")[3].find("a").text.split("\xa0\xa0\xa0")
    dateColumn.append(date)
    timeColumn.append(time[:-5])
    latituteColumn.append( e.select("td")[4].text.replace("\xa0","")+ e.select("td")[5].text.replace("\xa0",""))
    longitudeColumn.append( e.select("td")[6].text.replace("\xa0","")+ e.select("td")[7].text.replace("\xa0","")) 
    regionColumn.append( e.select("td")[11].text.replace("\xa0","").capitalize())
    
EarthDict={
    "Date":dateColumn,
    "Time":timeColumn,
    "Latitude":latituteColumn,
    "Longitude":longitudeColumn,
    "Region":regionColumn
}
display(pd.DataFrame.from_dict(EarthDict))

200
<Response [200]>



Date   Time Latitude Longitude                            Region
0   2019-09-06  12:56   36.78N   120.87W                Central california
1   2019-09-06  12:33   38.88N    26.27E  Near the coast of western turkey
2   2019-09-06  12:02    3.92N   126.50E       Kepulauan talaud, indonesia
3   2019-09-06  10:50   38.72N    26.56E  Near the coast of western turkey
4   2019-09-06  10:37   18.10S   167.63E                           Vanuatu
5   2019-09-06  10:32   13.87N   120.65E              Mindoro, philippines
6   2019-09-06  10:31   37.16N    30.68E                    Western turkey
7   2019-09-06  10:25   37.62N   118.86W      Long valley area, california
8   2019-09-06  10:09   21.22N    94.73E                           Myanmar
9   2019-09-06  09:12   17.72N    72.05W                      Haiti region
10  2019-09-06  09:00   36.52N    27.85E  Dodecanese is.-turkey border reg
11  2019-09-06  08:59   35.77N    31.70E                     Cyprus region
12  2019-09-06  08:49   52.79N   159.78E       Off east coast of kamchatka
13  2019-09-06  08:39   29.77S    71.53W          Offshore coquimbo, chile
14  2019-09-06  08:24   36.12N     7.68W               Strait of gibraltar
15  2019-09-06  08:21   29.73S    71.74W          Offshore coquimbo, chile
16  2019-09-06  08:05   42.80N    13.14E                     Central italy
17  2019-09-06  08:00   35.76N   117.56W               Southern california
18  2019-09-06  07:46    7.74S   118.95E                        Flores sea
19  2019-09-06  07:34   37.06N   121.10W                Central california
20  2019-09-06  07:28    3.23N   127.00E       Kepulauan talaud, indonesia
21  2019-09-06  07:25   28.70N   104.77E            Eastern sichuan, china
22  2019-09-06  06:49   35.92N   117.69W                Central california
23  2019-09-06  06:41   39.11N    48.40E                        Azerbaijan
24  2019-09-06  06:29   45.34N    26.21E                           Romania
25  2019-09-06  06:18   41.86N    14.88E                    Southern italy
26  2019-09-06  06:15    1.06N    30.29E        Lake albert region, uganda
27  2019-09-06  06:09   35.69N   117.55W               Southern california
28  2019-09-06  05:55   45.53N    26.48E                           Romania
29  2019-09-06  05:44   16.92S    71.33W                     Southern peru
30  2019-09-06  05:43   37.59N    20.74E                        Ionian sea
31  2019-09-06  05:40   26.40N    93.30E                      Assam, india
32  2019-09-06  05:27   36.41N    28.65E  Dodecanese is.-turkey border reg
33  2019-09-06  05:13   35.64N   117.61W               Southern california
34  2019-09-06  05:04   15.79S    71.75W                     Southern peru
35  2019-09-06  04:59   38.37N    30.91E                    Western turkey
36  2019-09-06  04:55   61.51N   149.90W                   Southern alaska
37  2019-09-06  04:34   32.82N   115.46W               Southern california
38  2019-09-06  04:26   16.26N   122.55E                Luzon, philippines
39  2019-09-06  03:50   44.57N    18.49E            Bosnia and herzegovina
40  2019-09-06  03:28   42.43N     6.90W                             Spain
41  2019-09-06  03:21   38.72N    38.15E                    Eastern turkey
42  2019-09-06  03:11   27.41S    70.24W                    Atacama, chile
43  2019-09-06  02:53   19.17N   155.49W          Island of hawaii, hawaii
44  2019-09-06  02:23   35.04N    24.54E                     Crete, greece
45  2019-09-06  01:54   22.83S    68.86W                Antofagasta, chile
46  2019-09-06  01:17    4.15S   152.65E        New britain region, p.n.g.
47  2019-09-06  01:15   10.19N    85.99W                        Costa rica
48  2019-09-06  01:07   35.19N    27.91E        Dodecanese islands, greece
49  2019-09-06  00:42    8.35N    83.02W                        Costa rica

#### Display the date, days, title, city, country of next 25 hackathon events as a Pandas dataframe table

In [274]:
# This is the url you will scrape in this exercise
url ='https://hackevents.co/search/anything/anywhere/anytime'


In [289]:
response = requests.get(url)
print(response.status_code)
print(response)
print()
soup = BeautifulSoup(response.text.encode('utf-8'))
with open("Output.txt", "w") as text_file:
    print(soup.encode("utf-8"), file=text_file)
table=soup.select('div[class="col-12 col-sm-7"]')

startDateList=[]
endDateList=[]
daysList=[]
titleList=[]
cityList=[]
countryList=[]
for element in table:
    startDate=element.select('p')[1].text.replace("\u2003Starts\u2003","")
    endDate=element.select('p')[2].text.replace("\u2003Ends\xa0\xa0\u2003","")
    startDateList.append(startDate)
    endDateList.append(endDate)
    titleList.append(element.find("h3").text)
    City,Country=element.find("p").text.split(",")
    cityList.append(City.replace("\u2003 ",""))
    countryList.append(Country[1:])
    
EventDict={
    "Date":startDateList,
    "Title":titleList,
    "City":cityList,
    "Country":countryList,
}
display(pd.DataFrame.from_dict(EventDict))

200
<Response [200]>



Date                    Title    City         Country
0   9/6/2019          TECHFEST MUNICH  Munich         Germany
1  1/31/2019  Galileo App Competition  Prague  Czech Republic

#### Count number of tweets by a given Twitter account.

You will need to include a ***try/except block*** for account names not found. 
<br>***Hint:*** the program should count the number of tweets for any provided account

In [24]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
urlDomain = "https://twitter.com/"

In [79]:
import os
from dotenv import load_dotenv
from os.path import join, dirname

# Put user name to check the tweets from here
#=============================================================================
urlUsername="neuer"
#=============================================================================

url=urlDomain+urlUsername
path= os.getcwd() + "\\token.env"
load_dotenv(dotenv_path=path,verbose=True)
username = os.getenv("USERNAME")
pwd=os.getenv("PWD")

response = requests.get(url, auth=(username,pwd))
print(response)
print(response.status_code)
print()
soup = BeautifulSoup(response.text.encode('utf-8'))
with open("Output.txt", "w") as text_file:
   print(soup.encode("utf-8"), file=text_file)
table=soup.select('a[class="ProfileNav-stat ProfileNav-stat--link u-borderUserColor u-textCenter js-tooltip js-nav"]')[0]
print("The number of tweet of user "+urlUsername+" is:")
#print(table[0].a['title'])
print(table['title'])

  

<Response [200]>
200

The number of tweet of user neuer is:
288.554 Tweets


#### Number of followers of a given twitter account

You will need to include a ***try/except block*** in case account/s name not found. 
<br>***Hint:*** the program should count the followers for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [57]:
followers=soup.select('[href$="/followers"]')[0].select('span')[2].text
try:
    print("The user "+ urlUsername + " has " + followers + " followers.")
except:
    print("The user "+ urlUsername + " has " + " no followers.")

The user neuer has 779 followers.


#### List all language names and number of related articles in the order they appear in wikipedia.org

In [4]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Main_Page'

In [11]:
#your code
response = requests.get(url)
print(response.status_code)
print(response)
print()
soup = BeautifulSoup(response.text.encode('utf-8'))
with open("Output.txt", "w") as text_file:
    print(soup.encode("utf-8"), file=text_file)
table=soup.select('li[id*="lang"]')
for element in table:
    print(element.text)


200
<Response [200]>

More than 1,000,000 articles: 
Deutsch
Español
Français
Italiano
Nederlands
日本語
Polski
Português
Русский
Svenska
Tiếng Việt
中文

More than 250,000 articles: 
العربية
Bahasa Indonesia
Bahasa Melayu
Български
Català
Čeština
Dansk
Esperanto
Euskara
فارسی
한국어
Magyar
Norsk Bokmål
Română
Srpski
Srpskohrvatski
Suomi
Türkçe
Українська

More than 50,000 articles: 
Bosanski
Eesti
Ελληνικά
English (Simple English)
Galego
עברית
Hrvatski
Latviešu
Lietuvių
മലയാളം
Norsk nynorsk
Slovenčina
Slovenščina
ไทย



#### A list with the different kind of datasets available in data.gov.uk 

In [91]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'

In [98]:
#your code
response = requests.get(url)
print(response.status_code)
print(response)
print()
soup = BeautifulSoup(response.text.encode('utf-8'))
with open("Output.txt", "w") as text_file:
    print(soup.encode("utf-8"), file=text_file)
table=soup.select('a[href*="/search?filters%5Btopic"]')
categories=[element.text for element in table]
print(categories)
    

200
<Response [200]>

['Business and economy', 'Crime and justice', 'Defence', 'Education', 'Environment', 'Government', 'Government spending', 'Health', 'Mapping', 'Society', 'Towns and cities', 'Transport']


#### Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [13]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'

In [74]:
#your code
response = requests.get(url)
print(response.status_code)
print(response)
print()
soup = BeautifulSoup(response.text.encode('utf-8'))
with open("Output.txt", "w") as text_file:
    print(soup.encode("utf-8"), file=text_file)
table=soup.select('table[class*="wikitable"]')[0]
row=table.select('tr')[1:]
result=[]
for element in row:
    result.append((element.select('td')[1].select('a[title]')[0].text))
langDict={'Position':list(range(1,11)),'Language':result[0:10]}  
langPd=pd.DataFrame.from_dict(langDict)
langPd=langPd.set_index('Position')
display(langPd)


200
<Response [200]>



Language
Position                 
1                Mandarin
2                 Spanish
3                 English
4                   Hindi
5                 Bengali
6              Portuguese
7                 Russian
8                Japanese
9         Western Punjabi
10                Marathi

### BONUS QUESTIONS

#### Scrape a certain number of tweets of a given Twitter account.

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
# your code

#### IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

In [ ]:
# your code

#### Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [ ]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [ ]:
#your code

#### Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [ ]:
#https://openweathermap.org/current
city = city=input('Enter the city:')
url = 'http://api.openweathermap.org/data/2.5/weather?'+'q='+city+'&APPID=b35975e18dc93725acb092f7272cc6b8&units=metric'

In [ ]:
# your code

#### Book name,price and stock availability as a pandas dataframe.

In [ ]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

In [ ]:
#your code